### PR TITLE
[#5324] Implemented atomic database operations function (4-2-stable)

### PR DESCRIPTION
**Pull request by korydraughn**
This new function executes multiple database operations under a single
transaction.

The following SQL operations are supported:
- insert
- update
- delete

The following tables are supported:
- r_data_main
- r_coll_main
- r_meta_main
- r_objt_metamap
- r_resc_main

This PR also includes work for a server-side only unit testing framework.

### CODE REVIEW

File: server/core/src/atomic_apply_database_operations.cpp [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+           coll_type        | character varying(250)  | default '0'::character varying
+           coll_info1       | character varying(2700) | default '0'::character varying
+           coll_info2       | character varying(2700) | default '0'::character varying
+           coll_expiry_ts   | character varying(32)   | 
+           r_comment        | character varying(1000) | 
+           create_ts        | character varying(32)   | 
+           modify_ts        | character varying(32)   |
+         */
+
+        // TODO Does the database schema protect the user from creating duplicates.
```
Comment (trel) [inline server/core/src/atomic_apply_database_operations.cpp:117]: yes, protected by the UNIQUE index(es):

```
Indexes:
    "idx_coll_main2" UNIQUE, btree (parent_coll_name, coll_name)
    "idx_coll_main3" UNIQUE, btree (coll_name)
    "idx_coll_main1" btree (coll_id)
```

File: server/core/src/atomic_apply_database_operations.cpp [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+                const auto data_status = row.value("data_status", "");
+                const auto data_checksum = row.value("data_checksum", "");
+                const auto data_expiry_ts = row.value("data_expiry_ts", "");
+                const auto data_map_id = row.value("data_map_id", 0);
+                const auto data_mode = row.value("data_mode", "");
+                const auto r_comment = row.value("r_comment", "");
+                const auto create_ts = row.value("create_ts", "");
+                const auto modify_ts = row.value("modify_ts", "");
+                const auto resc_hier = row.value("resc_hier", "");                          // TODO Should this be required?
+                const auto resc_id = row.value("resc_id", 0);                               // TODO Should this be required?
```
Comment (trel) [inline server/core/src/atomic_apply_database_operations.cpp:267]: this is the one we actually use now - not resc_name...

File: server/core/src/atomic_apply_database_operations.cpp [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+    } // update_rows
+} // anonymous namespace
+
+namespace irods::experimental
+{
+    auto atomic_apply_database_operations(RsComm& _comm, const nlohmann::json& _ops) -> int
+    {
+        auto [db_instance_name, db_conn] = ic::new_database_connection();
+
+        // The "&db_conn = db_conn" is required to avoid a bug with the C++ standard around structured bindings.
```
Comment (trel) [inline server/core/src/atomic_apply_database_operations.cpp:510]: &db_conn --> &dbc

**Comment by korydraughn** [comment]
How do we feel about the following JSON input structure for the **update** operation:
```javascript
{
    op_name: "update",
    table: "r_resc_main",

    // Allows multiple updates to be submitted for "r_resc_main".
    <WHAT_SHOULD_THIS_BE_CALLED???>: [
        {
            values: {
                "some_column_1": "",
                "some_column_2": "",
                "some_column_N": ""
            },

            // The list implies "AND".
            conditions: [
                {
                    column: "data_name",
                    operator: "like", // or "=", ">=", "<=", etc.
                    value: ["foo%"]
                },
                {
                    column: "data_size",
                    operator: "in",
                    value: ["3", "7"]
                }
            ]
        },
        {
            // A separate update operation for "r_resc_main" that can focus
            // on something completely different.
        }
    ]
}
```

**Comment by trel** [comment]
i like the structure - it mirrors the insert and deletes where they are consistent.

>     <WHAT_SHOULD_THIS_BE_CALLED???>: [



an sql statement consists of a command and clauses.

so i nominate `clauses`.

File: server/core/src/atomic_apply_database_operations.cpp [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+                });
+
+            // Join the conditions together.
+            std::ostringstream conditions_stream;
+            std::copy(std::begin(condition_clauses),
+                      std::end(condition_clauses),
+                      std::experimental::make_ostream_joiner(conditions_stream, " and "));
+
+            // Generate the final SQL.
+            const auto sql = fmt::format("delete from {} where {}", table_name, conditions_stream.str());
```
Comment (korydraughn) [inline server/core/src/atomic_apply_database_operations.cpp:551]: `table_name` needs to be capitalized here.

File: server/core/src/atomic_apply_database_operations.cpp [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+                    });
+
+                // Join the conditions together.
+                std::ostringstream conditions_stream;
+                std::copy(std::begin(condition_clauses),
+                          std::end(condition_clauses),
+                          std::experimental::make_ostream_joiner(conditions_stream, " and "));
+
+                // Generate the final SQL.
+                const auto sql = fmt::format("update {} set {} where {}", table_name, set_stream.str(), conditions_stream.str());
```
Comment (korydraughn) [inline server/core/src/atomic_apply_database_operations.cpp:643]: `table_name` needs to be capitalized here.

File: server/core/src/atomic_apply_database_operations.cpp [hunk 6]
```diff
@@ -0,0 +1,757 @@
...
+    auto atomic_apply_database_operations(RsComm& _comm, const nlohmann::json& _ops) -> int
+    {
+        rodsLog(LOG_NOTICE, "%s :: input = %s", __func__, _ops.dump(4).data());
+
+        std::string db_instance_name;
+        nanodbc::connection db_conn;
+        std::tie(db_instance_name, db_conn) = ic::new_database_connection();
+
+        // The "&dbc = db_conn" is required to avoid a bug with the C++ standard around structured bindings.
+        // See https://stackoverflow.com/a/46115028/5818611 for details.
```
Comment (korydraughn) [inline server/core/src/atomic_apply_database_operations.cpp:705]: This comment can be removed if structured binding aren't used to extract `db_instance_name` and `db_conn`.

File: server/core/src/atomic_apply_database_operations.cpp [hunk 7]
```diff
@@ -0,0 +1,757 @@
...
+                                       "values (nextval('R_OBJECTID'), ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)");
+#endif
+                const auto coll_id = row.at("coll_id").get<int>();                          // Cannot be null
+                const auto data_name = row.at("data_name").get<std::string>();              // Cannot be null
+                const auto data_repl_num = row.at("data_repl_num").get<int>();              // Cannot be null
+                const auto data_version = row.value("data_version", "0");
+                const auto data_type_name = row.at("data_type_name").get<std::string>();    // Cannot be null
+                const auto data_size = row.value("data_size", 0);                           // Cannot be null
+                const auto resc_group_name = row.value("resc_group_name", "");
+                const auto resc_name = row.at("resc_name").get<std::string>();              // Cannot be null
```
Comment (korydraughn) [inline server/core/src/atomic_apply_database_operations.cpp:332]: do not expose these.

File: server/core/src/atomic_apply_database_operations.cpp [hunk 8]
```diff
@@ -0,0 +1,757 @@
...
+                const auto data_is_dirty = row.value("data_is_dirty", 0);
+                const auto data_status = row.value("data_status", "");
+                const auto data_checksum = row.value("data_checksum", "");
+                const auto data_expiry_ts = row.value("data_expiry_ts", "");
+                const auto data_map_id = row.value("data_map_id", 0);
+                const auto data_mode = row.value("data_mode", "");
+                const auto r_comment = row.value("r_comment", "");
+                const auto create_ts = row.value("create_ts", "");
+                const auto modify_ts = row.value("modify_ts", "");
+                const auto resc_hier = row.value("resc_hier", "");                          // TODO Should this be required?
```
Comment (korydraughn) [inline server/core/src/atomic_apply_database_operations.cpp:345]: do not expose this.

File: server/core/src/atomic_apply_database_operations.cpp [hunk 9]
```diff
@@ -0,0 +1,757 @@
...
+                const auto data_owner_zone = row.at("data_owner_zone").get<std::string>();  // Cannot be null
+                const auto data_is_dirty = row.value("data_is_dirty", 0);
+                const auto data_status = row.value("data_status", "");
+                const auto data_checksum = row.value("data_checksum", "");
+                const auto data_expiry_ts = row.value("data_expiry_ts", "");
+                const auto data_map_id = row.value("data_map_id", 0);
+                const auto data_mode = row.value("data_mode", "");
+                const auto r_comment = row.value("r_comment", "");
+                const auto create_ts = row.value("create_ts", "");
+                const auto modify_ts = row.value("modify_ts", "");
```
Comment (korydraughn) [inline server/core/src/atomic_apply_database_operations.cpp:344]: default all timestamps to the time of the invocation.

File: server/core/src/atomic_apply_database_operations.cpp [hunk 10]
```diff
@@ -0,0 +1,757 @@
...
+    } // update_rows
+
+    auto throw_if_invalid_table_column(const std::string_view _column,
+                                       const std::string_view _table_name) -> void
+    {
+        const auto iter = std::find(std::begin(db_columns),
+                                    std::end(db_columns),
+                                    db_column_type{_table_name, _column});
+
+        if (iter == std::end(db_columns)) {
```
Comment (korydraughn) [inline server/core/src/atomic_apply_database_operations.cpp:673]: use `std::none_of`

File: server/core/src/atomic_apply_database_operations.cpp [hunk 11]
```diff
@@ -0,0 +1,757 @@
...
+            const auto msg = fmt::format("Unsupported conditional operator [{}].", _operator);
+            rodsLog(LOG_ERROR, "%s :: %s", __func__, msg.data());
+            THROW(SYS_INVALID_INPUT_PARAM, msg);
+        }
+    } // throw_if_invalid_conditional_operator
+} // anonymous namespace
+
+namespace irods::experimental
+{
+    auto atomic_apply_database_operations(RsComm& _comm, const nlohmann::json& _ops) -> int
```
Comment (korydraughn) [inline server/core/src/atomic_apply_database_operations.cpp:696]: remove the `_comm`. it isn't needed.

File: server/core/src/atomic_apply_database_operations.cpp [hunk 12]
```diff
@@ -19,6 +19,36 @@
...
+// TEAM DISCUSSION:
+// - Keep it a library (APIs can invoke this and guard this call with permission checks, etc.).
+// - Automated Ingest, NFSRODS, Change Log Listener would use this library.
+// - IDEA: Any API Plugin that invokes this library needs to provide a [cryptographic] access token in order for things to work.
+//   - How do we provide protection around this?
+//     - Could have a special group that can execute this.
+//     - Do we narrow it even more and put the secret in each server_config.json?
+//       - Scoped to a particular server.
+//       - The plugin specific configuration can contain a list of blessed users and groups, etc.
+//         - Example: {"user_1": "token_1", "user_2": "token_2", "group_1": "token_3"}
```
Comment (trel) [inline server/core/src/atomic_apply_database_operations.cpp:31]: so, this is an authenticated user who ALSO has this token?

or just a defined user in the system who uses this particular skeleton key to do ingest things, and might authenticate separately for other, more normal operations...

File: server/core/src/atomic_apply_database_operations.cpp [hunk 13]
```diff
@@ -0,0 +1,787 @@
...
+// iadmin addchildtoresc <parent> <child>
+// iadmin moveresc <resc> <from> <to>
+// iadmin mvresc <resc> <from> <to>
+// iadmin mr <resc> <from> <to>
+// iadmin mkparentresc <resc> <parent>
+// iadmin makeparentresc <resc> <parent>
+// iadmin mkpr <resc> <parent>
+// iadmin newparent <resc> <parent>
+//
+// **iadmin setparentresc <resc> <parent>**
```
Comment (trel) [inline server/core/src/atomic_apply_database_operations.cpp:46]: more consistent with `addchildtoresc` and `rmchildfromresc`...

would name parent first...

`iadmin setparentresc parent child`

also removes the need to check for empty string case.

File: server/core/src/atomic_apply_database_operations.cpp [hunk 14]
```diff
@@ -0,0 +1,787 @@
...
+            //     https://github.com/nanodbc/nanodbc/issues/56
+            //
+            for (auto&& row : _op.at(jp_values)) {
+#else
+            for (auto&& row : _op.at(jp_values)) {
+                nanodbc::prepare(stmt, "insert into R_COLL_MAIN ("
+                                       " coll_id, parent_coll_name, coll_name, coll_owner_name, coll_owner_zone, coll_map_id,"
+                                       " coll_inheritance, coll_type, coll_info1, coll_info2, coll_expiry_ts, r_comment,"
+                                       " create_ts, modify_ts) "
+                                       "values (nextval('R_OBJECTID'), ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)");
```
Comment (korydraughn) [inline server/core/src/atomic_apply_database_operations.cpp:251]: adjust the sequence number call based on the database instance.

File: server/core/src/atomic_apply_database_operations.cpp (hunk 13 above)
Comment (korydraughn) [reply server/core/src/atomic_apply_database_operations.cpp:48]: I'm more for specifying the `child` first in regards to `setparentresc`. I believe it is clearer and easier to remember. I don't think being consistent with `addchildtoresc` and `rmchildfromresc` helps the user very much.

I'm willing to discuss further if you're interested.

Comment (trel) [reply server/core/src/atomic_apply_database_operations.cpp:48]: further discussion at https://github.com/irods/irods/issues/5481

**Comment by korydraughn** [comment]
After some discussion, this enhancement will only be available in 4.3.0+.

See https://github.com/irods/irods/pull/5784 for more recent work.

File: server/core/src/atomic_apply_database_operations.cpp (hunk 12 above)
Comment (korydraughn) [reply server/core/src/atomic_apply_database_operations.cpp:31]: There is no authentication step for this function. This function is not exposed publicly. New RPC APIs will be able to use this for database operations.

Comment (trel) [reply server/core/src/atomic_apply_database_operations.cpp:31]: gotcha.

**Pull request closed without merging**